File: native/Sources/OpenVoiceFlow/LeaderboardAlias.swift
import Foundation

/// A readable, anonymous leaderboard alias. Generated aliases are one token
/// so they scan like handles rather than accidental first-and-last names.
enum LeaderboardAlias {
    private static let adjectives = [
        "Quiet", "Swift", "Calm", "Bright", "Steady", "Clever", "Brisk", "Gentle",
        "Sunny", "Nimble", "Bold", "Sharp", "Warm", "Cool", "Vivid", "Keen",
    ]
    private static let nouns = [
        "Falcon", "Otter", "Maple", "Comet", "Harbor", "Ember", "Willow", "Lynx",
        "Meadow", "Aspen", "Heron", "Cedar", "Ridge", "Sparrow", "Tundra", "Coral",
    ]

    static func make(adjective: String, noun: String, number: Int) -> String {
        "\(adjective)\(noun)\(number)"
    }

    static func random() -> String {
        make(
            adjective: adjectives.randomElement()!,
            noun: nouns.randomElement()!,
            number: Int.random(in: 1000...9999)
        )
    }

    /// Compact only names that exactly match the app's old generated format.
    /// User-chosen display names keep their original spacing.
    static func compactLegacyDefault(_ name: String) -> String {
        let parts = name.components(separatedBy: " ")
        guard parts.count == 3,
              adjectives.contains(parts[0]),
              nouns.contains(parts[1]),
              let number = Int(parts[2]),
              (10...99).contains(number),
              String(number) == parts[2]
        else { return name }

        return make(adjective: parts[0], noun: parts[1], number: number)
    }
}

File: api/_leaderboardAlias.js
const ADJECTIVES = new Set([
  "Quiet", "Swift", "Calm", "Bright", "Steady", "Clever", "Brisk", "Gentle",
  "Sunny", "Nimble", "Bold", "Sharp", "Warm", "Cool", "Vivid", "Keen",
]);
const NOUNS = new Set([
  "Falcon", "Otter", "Maple", "Comet", "Harbor", "Ember", "Willow", "Lynx",
  "Meadow", "Aspen", "Heron", "Cedar", "Ridge", "Sparrow", "Tundra", "Coral",
]);

/// Mirrors LeaderboardAlias.compactLegacyDefault (native/Sources/OpenVoiceFlow/LeaderboardAlias.swift).
/// Devices that haven't relaunched since the alias format changed still have
/// "Adjective Noun NN" stored server-side; compact it here so every viewer
/// sees the current no-space format regardless of which client wrote it.
/// User-chosen display names never match this exact shape, so they pass through untouched.
export function compactLegacyDefault(name) {
  const parts = name.split(" ");
  if (parts.length !== 3) return name;
  const [adjective, noun, numberText] = parts;
  if (!ADJECTIVES.has(adjective) || !NOUNS.has(noun)) return name;
  const number = Number(numberText);
  if (!/^\d{2}$/.test(numberText) || number < 10 || number > 99) return name;
  return `${adjective}${noun}${number}`;
}
